### DISCUSSION TITLE
Question about anvil internals

### DISCUSSION THREAD

**Discussion by aathan** [foundry-rs/foundry#5387] [Help]
See https://github.com/foundry-rs/foundry/pull/5378#issuecomment-1635044052 (copied here):

Two things, so I can contribute more effectively:

I noticed https://github.com/foundry-rs/foundry/commit/a516b21ae627d8d727adc3839c358c6a6e3b5c3c env.cfg is used vs env.read().cfg -- when is it ok not to lock env for reading?

Also, can you explain why the u64 max is in fact correct vs the U256::max_value()? Why should there be a problem converting a U256 max to its equiv revm type?

**Comment by onbjerg** [comment]
For the commit referenced, the env is not wrapped in a lockable type.

For question two, can you please specify what you are talking about? While some types are u256 according to the Ethereum spec, there are some implied limits that make it possible to pack the values in u64's instead. For example, a transaction on mainnet cannot go over 30m gas, which is more than comfortably representable by u64. For value transfers, technically the field is u256, but again, it is inconceivable that anyone would (currently) transfer that much eth, so it can be represented by a smaller type.

**Reply by aathan** [reply]
For (1): I guess knowing when it is or isn't wrapped is a matter of being familiar with the code? What's my clue?

For (2): See https://github.com/foundry-rs/foundry/pull/5378/commits/b3eecba41702ddeb9b6351bb627c54b68e231b55 and https://github.com/foundry-rs/foundry/pull/5378#issuecomment-1635578677

... the title says "... avoid panic ...", mattse's comment/response seems to imply just some type conversion issues, the code references a call into `u256_to_ru256()` for which it seems completely natural to provide `U256::max_value()`

So there appears to be some mix of semantics, type conversion issues that are exogenous to the EVM / more related to implementation choices/issues of the libraries/tools being used, and lore/hidden knowledge going on here.

Thus, why I asked. Just trying to learn how to get hints from the code I'm looking at so I can contribute better.

I will say however, that if there is indeed a panic converting a U256:0xF* value to the maximum allowed gas and the panic happens in revm, and the EVM limit is anyhow less than u64::max, that seems like a "bug" or at least, like a strange choice. Why panic on U256::max but not u64::max if both are above the EVM standard's limit? Might be worth an upstream bug or highlight?


Thanks for the education.
